### DISCUSSION TITLE
Add non-member type_name function

### DISCUSSION THREAD

**Discussion by SamVanheer** [nlohmann/json#3960] [Ideas]
There is currently no way to get the JSON type name for a `value_t` directly. You have to construct a json value to get it: `json(json::value_t::object).type_name()`

It would be useful to have a function to get this directly.

**Comment by nlohmann** [comment]
You could implement this with a switch over the enum `json::value_t`. What is your use case that you need these names?

**Reply by SamVanheer** [reply]
I'm generating a JSON Schema using [json schema validator](https://github.com/pboettch/json-schema-validator) and i have a set of objects that describe what type they are by returning one of the enumeration values. This is so that i can automate part of the schema generation and do trivial rejection at runtime when loading JSON data.

This is a simplified version of what i've built:
```cpp
class GameConfigSection
{
protected:
	GameConfigSection() = default;

public:
	GameConfigSection(const GameConfigSection&) = delete;
	GameConfigSection& operator=(const GameConfigSection&) = delete;
	virtual ~GameConfigSection() = default;

	virtual std::string_view GetName() const = 0;
	virtual json::value_t GetType() const = 0;
	virtual std::string GetSchema() const = 0;
};

// Constructing the schema for each section:
std::string buffer;

bool first = true;

for (const auto& section : m_Sections)
{
	if (first)
	{
		first = false;
	}
	else
	{
		buffer += ',';
	}

	const auto schema = section->GetSchema();

	buffer += fmt::format(R"(
"{0}": {{
	"title": "{0} Section",
	"type": "{1}"{2}
	{3}
}})",
		section->GetName(), json(section->GetType()).type_name(), schema.empty() ? ""sv : ","sv, schema);
}

// At parse time:
if (sectionObj->GetType() != section.type())
{
	continue;
}
```

Using an enumeration is more robust than returning the string name.
